libminifi/src/io/ZlibStream.cpp

```diff
@@ -0,0 +1,202 @@
+  }
+
+  state_ = ZlibStreamState::INITIALIZED;
+}
+
+ZlibCompressStream::~ZlibCompressStream() {
+  if (state_ != ZlibStreamState::UNINITIALIZED) {
+    deflateEnd(&strm_);
+  }
+}
+
+int ZlibCompressStream::writeData(uint8_t* value, int size) {
+  if (state_ != ZlibStreamState::INITIALIZED) {
+    logger_->log_error("writeData called in invalid ZlibCompressStream state, state is %hhu", state_);
+    return -1;
+  }
+
+  strm_.next_in = value;
+  strm_.avail_in = size;
+
+  /*
+   * deflate consumes all input data it can (i.e. if it has enough output buffer it never leaves input data unconsumed)
+   * and fills the output buffer to the brim every time it can. This means that the proper way to use deflate is to
+   * watch avail_out: once deflate does not have to fill it to the brim, then it has consumed all data we have provided
+   * to it, and does not need more output buffer for the time being.
+   * When we have no more input data, we must call deflate with Z_FINISH to make it empty its internal buffers and
+   * close the compressed stream.
+   */
+  do {
+    logger_->log_trace("writeData has %u B of input data left", strm_.avail_in);
+
+    strm_.next_out = outputBuffer_.data();
+    strm_.avail_out = outputBuffer_.size();
+
+    int flush = value == nullptr ? Z_FINISH : Z_NO_FLUSH;
+    logger_->log_trace("calling deflate with flush %d", flush);
+
+    int ret = deflate(&strm_, flush);
+    if (ret == Z_STREAM_ERROR) {
+      logger_->log_error("deflate failed, error code: %d", ret);
+      state_ = ZlibStreamState::ERRORED;
+      return -1;
+    }
+    int output_size = outputBuffer_.size() - strm_.avail_out;
+    logger_->log_trace("deflate produced %d B of output data", output_size);
+    if (BaseStream::writeData(outputBuffer_.data(), output_size) != output_size) {
+      logger_->log_error("Failed to write to underlying stream");
+      state_ = ZlibStreamState::ERRORED;
+      return -1;
+    }
+  } while (strm_.avail_out == 0);
+
+  return size;
+}
+
+void ZlibCompressStream::closeStream() {
+  if (state_ == ZlibStreamState::INITIALIZED) {
+    if (writeData(nullptr, 0U) == 0) {
+      state_ = ZlibStreamState::FINISHED;
+    }
+  }
+}
+
+/* ZlibDecompressStream */
+
+ZlibDecompressStream::ZlibDecompressStream(bool gzip)
+  : ZlibDecompressStream(this, gzip) {
+}
+
+ZlibDecompressStream::ZlibDecompressStream(DataStream* other, bool gzip)
+    : ZlibBaseStream(other)
+    , logger_(logging::LoggerFactory<ZlibDecompressStream>::getLogger()) {
+  int ret = inflateInit2(&strm_, 15 + (gzip ? 16 : 0) /* windowBits */);
+  if (ret != Z_OK) {
+    logger_->log_error("Failed to initialize z_stream with inflateInit2, error code: %d", ret);
+    throw Exception(ExceptionType::GENERAL_EXCEPTION, "zlib inflateInit2 failed");
+  }
+
+  state_ = ZlibStreamState::INITIALIZED;
+}
+
+ZlibDecompressStream::~ZlibDecompressStream() {
+  if (state_ != ZlibStreamState::UNINITIALIZED) {
```

Comment:
The same here, uninited state can't happen, although I wonder if inflateEnd should be called in case decompress failed and we have an error state.

Reply:
inflateEnd frees the resources allocated by zlib, so if we called inflateInit and it succeeded we must call inflateEnd, even if there as an error.

Reply:
Okay, thanks, that makes sense.

Reply:
Would it make sense to make `state_ != UNINITIALIZED` a postcondition of the ctor and remove the UNINITIALIZED state? It seems to be already the case, so we could remove all possibility of related dead code by removing the enum value. We don't need to prepare for any usage of an object, including destruction, during object creation.

Reply:
Not if you ever want to make this object movable: in that case you would have to set the moved-from object to UNINITIALIZED to prevent it from freeing the z_stream.